### ISSUE TITLE
Integrate historic data stored on device

### ISSUE THREAD

**Issue by Luminger** [Bluetooth-Devices/thermopro-ble#9]
**Is your feature request related to a problem? Please describe.**
The official Thermopro app is able to query the historic data for (at least) the TP357. It would be really nice if this integration would be able to actively query the stored data and populate this historic data for HA to use.

**Describe the solution you'd like**
After some digging I surfaced https://github.com/pasky/tp357, where @pasky (kudos!) documented and implemented the querying. I'd assume you're able to integrate the method used, however my understanding of the HA API and how the recorder works is lacking.

**Additional context**
I'd give it a shot myself but I'm currently too busy with other things. As mentioned, I also haven't checked if the API even allows for historic data to be send to a recorder... If the interface is there it would be really neat to have this integrated.

**Comment by tbm** [comment]
I'm also interested in this feature. @Luminger do you have time to work on this? (Unfortunately I lack the skills)

**Comment by bluetoothbot** [comment]
## Design proposal — historic data fetch over GATT

Picking this up as a design spike. Want to land the protocol contract and library
shape first so a contributor with a TP357 in hand can do the packet-capture work
without re-deriving the API surface. Posting here for discussion before opening a
PR — feedback welcome on naming and scope.

### Context

- The library today is **passive only**: `ThermoProBluetoothDeviceData` decodes
  advertisements, never connects.
- The one exception is `ThermoProDevice.set_datetime()` (`device.py`), which
  writes the current wall-clock to `00010203-0405-0607-0809-0a0b0c0d2b11`. That
  same UUID is the command channel @pasky reverse-engineered for history reads.
- Polling scaffolding is landing in #165 (`poll_needed` / `async_poll`) but only
  **wakes** silent devices — it does not pull stored history.

### Protocol (TP357, from pasky/tp357 + on-wire notes)

**Characteristics (unchanged from `set_datetime`):**

| Direction | UUID                                          | Use                |
| --------- | --------------------------------------------- | ------------------ |
| Write     | `00010203-0405-0607-0809-0a0b0c0d2b11`        | Commands           |
| Notify    | `00010203-0405-0607-0809-0a0b0c0d2b10`        | Live + history     |

**Frame discrimination** is by the first byte of each notify payload:

| Byte 0 | Meaning                              |
| ------ | ------------------------------------ |
| `0xA5` | Datetime ack (mirror of write frame) |
| `0xC2` | Live current-reading push            |
| `0xCx` | History block (subtype TBD)          |

A history pull needs: (a) a fresh `set_datetime` so the device's clock matches
host time (otherwise samples come back with wall-clock drift), (b) a "dump
history" command frame, (c) a stream of fixed-width sample frames over notify,
terminated by an end-of-stream marker.

The exact request/response bytes need to be confirmed with hardware — see
"Open questions" below.

### Proposed API

Keep it on the existing `ThermoProDevice` (no new class), mirroring how
`set_datetime` is structured:

```python
@dataclass(frozen=True, slots=True)
class HistoricalReading:
    timestamp: datetime          # UTC, derived from device clock + offset
    temperature: float           # °C
    humidity: float | None       # %, None if device only logs temp

class ThermoProDevice:
    HISTORY_SUPPORTED_MODELS: frozenset[str] = frozenset({"TP357", "TP357S"})
    # extend as captures land

    @classmethod
    def supports_history(cls, model_or_name: str) -> bool: ...

    async def fetch_history(
        self,
        *,
        since: datetime | None = None,
        sync_clock: bool = True,
    ) -> list[HistoricalReading]:
        """Pull stored samples from the device.

        Connects, optionally syncs the clock so timestamps line up,
        sends the history-dump command, collects notifications until
        the end-of-stream marker, then disconnects.
        """
```

**Why a list and not an async iterator?** Devices store on the order of
hours/days of samples at minute granularity — a few KB of payload. The
streaming bit happens internally over notifications; exposing it as an
iterator would force every caller to hold the connection open while it
processes. A flat list keeps the connection short and matches how
`bluetooth_sensor_state_data` already shapes consumer APIs.

### Capability flag (works with #168)

If the model registry from #168 lands first, this can be a
`CAP_FETCH_HISTORY = "fetch_history"` flag on each `ThermoProModel`:

```python
ThermoProModel(
    name="TP357",
    family="TP35x",
    capabilities=(CAP_SET_DATETIME, CAP_FETCH_HISTORY),
)
```

Then `supports_history()` becomes a thin wrapper over
`registry.has_capability(model, CAP_FETCH_HISTORY)` — same pattern as the
datetime gate.

### Home Assistant integration shape

The HA recorder accepts backfilled statistics via
`async_import_statistics` (long-term stats) and `async_add_external_statistics`.
That happens **downstream** of this library. The library's contract should be:

1. Return `list[HistoricalReading]` sorted ascending by timestamp.
2. Timestamps are timezone-aware UTC.
3. Caller is responsible for deduplication against existing recorder rows
   (the device has no notion of "what you've already seen").

The HA-side PR is out of scope here — this issue tracks the library API only.

### Open questions (need TP357 hardware to answer)

1. **Command frame bytes**: what does the host write to ask for history?
   `0xA5 … 0x5A` framing like `set_datetime`, or different?
2. **Sample frame format**: how many samples per notify, what's the timestamp
   encoding (LE16 minute-offset from the dump-start, or absolute)?
3. **End-of-stream marker**: explicit terminator byte, length-prefixed, or just
   "no more notifications for N ms"?
4. **Storage depth & resolution**: how far back does the device keep, at what
   sample interval? The HA app advertises ~24h for TP357.
5. **Other models**: TP358/TP358S/TP393 share the datetime UUID — do they also
   expose the same history command? Each needs a capture before being added to
   `HISTORY_SUPPORTED_MODELS`.

### Suggested path forward

1. Land the data class + the `supports_history` gate + the method **stub** that
   raises `NotImplementedError`. Zero risk, makes the contract concrete,
   downstream HA work can start in parallel.
2. Open a separate PR per model once captures arrive, each one filling in the
   protocol bytes for that family.

That keeps each PR small and reviewable, and avoids landing a half-decoder that
silently produces wrong timestamps for unverified models.

---

Tagging this so it doesn't get lost. Happy to do step 1 (the contract-only PR)
once the API shape above gets a thumbs-up. The reverse-engineering work needs
someone with a device to confirm the bytes — I can't fly that part blind.
